inception-external-search-core/src/main/java/de/tudarmstadt/ukp/inception/externalsearch/ExternalSearchHighlight.java

```diff
@@ -0,0 +1,66 @@
+/*
+ * Copyright 2019
+ * Ubiquitous Knowledge Processing (UKP) Lab
+ * Technische Universität Darmstadt
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package de.tudarmstadt.ukp.inception.externalsearch;
+
+import java.io.IOException;
+
+import java.io.Serializable;
+import java.util.ArrayList;
+import java.util.List;
+
+import com.google.common.base.Optional;
+import de.tudarmstadt.ukp.inception.support.annotation.OffsetSpan;
+
+/**
+ * This class represents a highlight from external search result.
+ * A highlight is a section of text containing occurrence(s) of keywords.
+ * The offsets tells where those keywords are within the original text document.
+ */
+public class ExternalSearchHighlight implements Serializable
+{
+
+    private static final long serialVersionUID = 4044755133816131842L;
+
+    private Optional<String> highlight;
+
+    private List<OffsetSpan> offsets;
+
+    public ExternalSearchHighlight(List<OffsetSpan> aOffsets) throws IOException
+    {
+        if (aOffsets.size() < 0) {
+            throw new IOException("Offsets must not be empty");
+        }
+        offsets = new ArrayList<>(aOffsets);
+
+    }
+
+    public ExternalSearchHighlight(String aHighlight, List<OffsetSpan> aOffsets) throws IOException
+    {
+        this(aOffsets);
+        highlight = Optional.fromNullable(aHighlight);
```

Comment:
If they are calling this constructor, `aHighlight` should be non-null, so you can use `Optional.of()`

Reply:
Optional is not serializable and shouldn't be used in member fields of serializable objects.

Reply:
Ah, good catch.  @shinyichen , just for future reference on other projects, even if we were using Guava, we'd want to store the internal field as the plain type and use `Optional.of` on the accessors. (I've gotten spoiled but not writing a value class without http://immutables.github.io/ for years :-) ).

Reply:
It's something that Wicket would badly choke on because it uses serialization to persists page state.

Reply:
I *love* Immutables and wouldn't touch Lombok with a ten-foot pole. :-) .   There are a few major differences:

* Lombok does a kitchen sink of different things which, when used heavily, can make the code difficult to understand for Java developers not familiar with the library.  Immutables does one thing: given an annotated interface for a value type, produce the "obvious" immutable or mutable implementation.
 
* More importantly, Lombok works by modifying the Java compiler itself (via a plugin, I think) and using internal and non-guaranteed APIs to inject code directly into annotated classes in-memory at compile-time.  Immutables, on the other hand, does things "the right way":  it runs as an annotation processor during compile time and then generates source code files on disk for the implementations of the interfaces which are then compiled (`javac` does multiple compilation passes when annotation processors are in use).  You can inspect these files in your IDE; they are clean and readable Java with no magic.  (The files are generated under `target/generated_sources`, so they are typically excluded from source control).  

At ISI, we mostly have to use Python ( :-( ), but at my previous job we used Immutables whenever possible. It's most notable benefits:
* it saves a huge amount of boilerplate.
* it avoids bugs like forgetting to update `equals` and `hashCode` when a new field is added.
* it made it easy to use the "immutable value class with builder + validation at initialization" pattern; using this whenever possible in our code cut down on bugs massively. When there was a bug, it was usually easy to find.
* it improved readability.  Once the team gets used to reading the fairly straightforward annotations, an Immutables-annotated interface because a very clear, compact, declarative summary of the behavior and constraints on a class.  Additionally, creating classes is now so quick and easy that people are more likely to break large complex classes into more closely-related sub-parts, etc. 

The only significant downsides are:
* most IDEs (e.g. IntelliJ) don't have annotation processing enabled by default, so it adds a trip to the preferences page during new developer setup for a project.
* Compilation errors in the annotated interface will block generation of the implementation class, leading to cascading compilation errors.  Once you know to look for this it is trivial to say "Oh, my `FooValue` interface has a syntax error, so I should look at the errors for that file, fix it, and ignore all the errors about `ImmutableFooValue` or `ModifiableFooValue` being missing.  But it can confused new users.  
* While not required, using a builder pattern and/or static factory methods instead of public constructors is a very natural and common with Immutables. This is unfamiliar to some developers.

Reply:
Removed Optional.